scipy/signal/_splines.py

```diff
@@ -0,0 +1,162 @@
+
+import numpy as np
+
+from ._arraytools import axis_slice, axis_reverse
+from ._signaltools import lfilter, sosfilt
+from ._spline import symiirorder1_ic, symiirorder2_ic_fwd, symiirorder2_ic_bwd
+
+__all__ = ['symiirorder1', 'symiirorder2']
+
+
+def symiirorder1(signal, c0, z1, precision=-1.0):
+    """
+    Implement a smoothing IIR filter with mirror-symmetric boundary conditions
+    using a cascade of first-order sections.  The second section uses a
+    reversed sequence.  This implements a system with the following
+    transfer function and mirror-symmetric boundary conditions::
+
+                           c0
+           H(z) = ---------------------
+                   (1-z1/z) (1 - z1 z)
+
+    The resulting signal will have mirror symmetric boundary conditions
+    as well.
+
+    Parameters
+    ----------
+    input : ndarray
+        The input signal. If 2D, then the filter will be applied in a batched
+        fashion across the last axis.
+    c0, z1 : scalar
+        Parameters in the transfer function.
+    precision :
+        Specifies the precision for calculating initial conditions
+        of the recursive filter based on mirror-symmetric input.
+
+    Returns
+    -------
+    output : ndarray
+        The filtered signal.
+    """
+    if np.abs(z1) >= 1:
+        raise ValueError('|z1| must be less than 1.0')
+
+    if signal.ndim > 2:
+        raise ValueError('Input must be 1D or 2D')
+
+    squeeze_dim = False
+    if signal.ndim == 1:
+        signal = signal[None, :]
+        squeeze_dim = True
+
+    if np.issubdtype(signal.dtype, np.integer):
+        signal = signal.astype(np.promote_types(signal.dtype, np.float32))
```

Comment:
So this relies on numpy propotining `(int, float32) -> float64`, it seems (which works on both numpy < 2 and 2.0.0rc1). Am I missing a reason to spell it like this instead of just `signal.astype(float)`?

Reply:
Is `float` an alias to the smallest floating type that can represent the values?

Reply:
No, `s.astype(float)` is an alias to `s.astype(np.float64)`. Which is what the `promote_type` stanza above resolves to:

```
In [1]: import numpy as np

In [2]: s = np.where(np.arange(100) %2, -1, 1)

In [3]: s.dtype
Out[3]: dtype('int64')

In [4]: np.promote_types(s.dtype, np.float32)
Out[4]: dtype('float64')
```

Which is correct from the backwards compat POV, since

```
In [6]: symiirorder1(s, 0.5, 0.5).dtype
Out[6]: dtype('float64')

In [7]: import scipy

In [8]: scipy.__version__
Out[8]: '1.7.1'

```